### PR TITLE
Creating setShadow to TextView (to be used from code)

### PR DESCRIPTION
**Pull request by anderbytes**
I wasn't able to test it as I didn't succeed on linking my Android Studio Kotlin project to locally implement the PhotoEditor Intellij project as a dependency on my Android app (sadly), but I'd like to propose this change as it seems simple (as others that may contribute in the future) related to TextStyling

Please test it or fix it if I forgot anything  ;-)

### CODE REVIEW

**Comment by burhanrashid52** [comment]
Please test the changes before creating a PR. This will avoid going back and forth if something get broken.
Also please add the tests for those changes to make sure someone doesn't break it in future.

**Comment by anderbytes** [comment]
I will try again to link the local project to my main project where I call the PhotoEditor class, so I can do some local testing. But as I don't "use" your sample editor, I will only do some testing with the methods (old and new).

What do you mean with "add those tests" ? Something related to Intellij JUNIT plugin? I'm pretty novice in this so I've never really used "test feature" of the IDE.

**Comment by burhanrashid52** [comment]
> I will try again to link the local project to my main project where I call the PhotoEditor class, so I can do some local testing. But as I don't "use" your sample editor, I will only do some testing with the methods (old and new).

If you can import as module and it should work.

> What do you mean with "add those tests" ? Something related to Intellij JUNIT plugin? I'm pretty novice in this so I've never really used "test feature" of the IDE.

Yes, Junit test or UI test to make sure the the text changes are applied to the textview


> If I ever manage to do it right, I'll be more than happy to continue collaborating with what I can, even if my main language being Kotlin
Thank you :)

**Comment by anderbytes** [comment]
@burhanrashid52 thanks for the module import tip. 

I did some testing here in my app and it worked perfectly! but I am clueless about using JUnit, even after seeing your PinchTestHelper in "src/androidTest", I can't figure how to properly test this new TextView method. As I mentioned... never ever worked with automated tests before.

Would you please create a test about a textview modification so I can learn it? You could even do one related to a feature that your app already does (ex: TextView font size change), and I'd understand it and create my own test right after, related to Shadow.

If everything goes well, I'll be soon adding more improvements, I could even be commiting a few more changes to TextView before you change from 1.1.4 to 1.1.5

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java [hunk 1]
```diff
@@ -32,6 +32,24 @@ public void withTextSize(@NonNull float size) {
...
 
+    /**
+     * Set this textShadow style
+     *
+     * @param radius Radius of the shadow to apply on text
+     * @param dx Horizontal distance of the shadow
+     * @param dy Vertical distance of the shadow
+     * @param color Color of the shadow
+     */
+    public void withTextShadow(@NonNull float radius, float dx, float dy, int color) {
```
Comment (burhanrashid52) [inline photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java:43]: Please move these 4 parameter to `TextShadow` object and put them in map with `TextStyle.SHADOW` key ?
`TextShadow` does not need to have setters they can be created using constructor only.

**Comment by burhanrashid52** [comment]
UI test for text is tricky. You can post the screenshot here to give an idea how it looks. Because the only way I see is to test using  [TextStyleBuilderTest](https://github.com/burhanrashid52/PhotoEditor/blob/master/photoeditor/src/test/java/ja/burhanrashid52/photoeditor/TextStyleBuilderTest.java) and check weather it's setting property on TextView properly.

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java (hunk 1 above)
Comment (anderbytes) [reply photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java]: Just did it, committing now. Hope I got it right in Java.

Also used the commit to do some little bumping in dependencies versions.

**Comment by anderbytes** [comment]
> Uploaded here some screenshots of my testing.
> 
> UI test for text is tricky. You can post the screenshot here to give an idea how it looks. Because the only way I see is to test using [TextStyleBuilderTest](https://github.com/burhanrashid52/PhotoEditor/blob/master/photoeditor/src/test/java/ja/burhanrashid52/photoeditor/TextStyleBuilderTest.java) and check weather it's setting property on TextView properly.

blackShadow - code
blackShadow - normal
blackShadow - zoom
noShadow - code
noShadow - normal
noShadow - zoom
redShadow - code
redShadow - normal
redShadow - zoom

**Comment by anderbytes** [comment]
ps: radius in the "red" test was actually 16F (greater than black one), I just took the screenshot of the code before changing that from 7F.

**Comment by anderbytes** [comment]
Hi @burhanrashid52 something went wrong ? I'm seeing some error above, but it's too generic and says something about a DEX file , which I don't know about.

I could remove the "version bumps" I did on the PR, would it help?

**Comment by burhanrashid52** [comment]
It failing because of 64K method limit.
```
Task :app:mergeDexDebugAndroidTest FAILED
ERROR:: D8: Cannot fit requested classes in a single dex file (# methods: 90699 > 65536)
com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
The number of method references in a .dex file cannot exceed 64K.
```
Learn how to resolve this issue at https://developer.android.com/tools/building/multidex.html

**Comment by anderbytes** [comment]
> 
> 
> It failing because of 64K method limit.
> 
> ```
> Task :app:mergeDexDebugAndroidTest FAILED
> ERROR:: D8: Cannot fit requested classes in a single dex file (# methods: 90699 > 65536)
> com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
> The number of method references in a .dex file cannot exceed 64K.
> ```
> 
> Learn how to resolve this issue at https://developer.android.com/tools/building/multidex.html



> 
> 
> It failing because of 64K method limit.
> 
> ```
> Task :app:mergeDexDebugAndroidTest FAILED
> ERROR:: D8: Cannot fit requested classes in a single dex file (# methods: 90699 > 65536)
> com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
> The number of method references in a .dex file cannot exceed 64K.
> ```
> 
> Learn how to resolve this issue at https://developer.android.com/tools/building/multidex.html

Is that "excess" related to the little modification I made?? That can't be real. 
This it getting real real hard you know. I'm justing adding a new "withTextShadow" property.

**Comment by burhanrashid52** [comment]
Updating the libraries might increase the methods count. Can you revert it back and check in your local ?

**Comment by anderbytes** [comment]
Just reverted the latest commit and re-inserted the actual TextShadow code.

If still too many methods, I could also try to revert the first commit (https://github.com/burhanrashid52/PhotoEditor/pull/370/commits/d20ed83d3056180795a879a076e8ac0297f08c08) , that also had some version bumps.

**Comment by burhanrashid52** [comment]
LGTM

**Review by burhanrashid52 — APPROVED** [review]

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java [hunk 2]
```diff
@@ -145,6 +164,10 @@ protected void applyTextSize(TextView textView, float size) {
         textView.setTextSize(size);
     }
 
+    protected void applyTextShadow(TextView textView, float radius, float dx, float dy, int color) {
```
Comment (burhanrashid52) [inline photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java:167]: You can replace this method with [this](https://github.com/burhanrashid52/PhotoEditor/blob/master/photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.java#L227). And just past TextShadow from in line 118.

You can also remove `TextShadow` inside the `TextShadowBuilder`

You need update with the latest master to this method.

**Pull request merged**